File: Homework/Son/W1/304-range-sum-query-2d-immutable.js
class NumMatrix {
    constructor(matrix){
        this.matrix = matrix;
        this.cache = [];
        this.preCompute();
    }

    preCompute(){
        for (let row = 0; row < this.matrix.length; row++) {
            this.cache[row] = [];
            for (let col = 0; col < this.matrix[0].length; col++) {
                let prevCol = col > 0 ? this.cache[row][col-1] : 0;
                let prevRow = row > 0 ? this.cache[row-1][col] : 0;
                let prevCache = row > 0 && col > 0 ? this.cache[row-1][col-1] : 0;
                let sum = this.matrix[row][col] + prevCol + prevRow - prevCache;
                this.cache[row].push(sum)
            }
        }
    }

    sumRegion(row1, col1, row2, col2){

        let totalArea = this.cache[row2][col2];
        let topArea = row1 > 0 ? this.cache[row1-1][col2] : 0;
        let sideArea = col1 > 0 ? this.cache[row2][col1-1] : 0;
        let cornerArea = row1 > 0 && col1 > 0 ? this.cache[row1-1][col1-1] : 0;

        return totalArea - topArea - sideArea + cornerArea;

    }
}

// https://leetcode.com/problems/range-sum-query-2d-immutable/

File: Homework/Ngan_Luu/304-range-sum-query-2d-immutable.py
#https://leetcode.com/problems/range-sum-query-2d-immutable/
class NumMatrix:
    def __init__(self, matrix: List[List[int]]):
        m = len(matrix)
        n = len(matrix[0])
        self.prefixSum = [[0]*(n+1) for i in range(m+1)]
        for i in range(1,m+1):
            for j in range(1,n+1):
                self.prefixSum[i][j]=self.prefixSum[i-1][j]+self.prefixSum[i][j-1]-self.prefixSum[i-1][j-1]+matrix[i-1][j-1]
    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        return self.prefixSum[row2+1][col2+1]-self.prefixSum[row1][col2+1]-self.prefixSum[row2+1][col1]+self.prefixSum[row1][col1]